extensions/positron-assistant/src/tools/projectTreeTool.ts

```diff
@@ -0,0 +1,287 @@
+	filterResults?: boolean;
+	maxFiles?: number;
+}
+
+export const ProjectTreeTool = vscode.lm.registerTool<ProjectTreeInput>(PositronAssistantToolName.ProjectTree, {
+	prepareInvocation: async (_options, _token) => {
+		return {
+			// The message shown when the code is actually executing.
+			// Positron appends '...' to this message.
+			invocationMessage: vscode.l10n.t('Constructing project tree'),
+			pastTenseMessage: vscode.l10n.t('Constructed project tree'),
+		};
+	},
+	/**
+	 * Called to get the project tree.
+	 * @param options The options for the tool invocation.
+	 * @param token The cancellation token.
+	 * @returns A vscode.LanguageModelToolResult containing the project tree.
+	 */
+	invoke: async (options, token) => {
+		const workspaceFolders = vscode.workspace.workspaceFolders;
+		if (!workspaceFolders || workspaceFolders.length === 0) {
+			throw new Error(`Can't construct project tree because no workspace folders are open. Open a workspace folder before using this tool.`);
+		}
+
+		log.debug(`[${PositronAssistantToolName.ProjectTree}] Constructing project tree for ${workspaceFolders.length} workspace folders...`);
+
+		// Parse the input options
+		const { include, exclude, replaceDefaultExcludes, excludeSettings, ignoreFiles, filterResults, maxFiles } = options.input;
+
+		log.trace(`[${PositronAssistantToolName.ProjectTree}] Invoked with options: ${JSON.stringify(options.input, null, 2)}`);
+
+		const filePatterns = include ?? DEFAULT_INCLUDE_PATTERNS;
+		const excludePatterns = exclude ?? [];
+		const filterResultsEnabled = filterResults ?? DEFAULT_FILTER_RESULTS;
+		const filesLimit = maxFiles ?? DEFAULT_MAX_FILES;
+
+		let findOptions: vscode.FindFiles2Options;
+		if (filterResultsEnabled) {
+			findOptions = {
+				exclude: replaceDefaultExcludes ? excludePatterns : [...DEFAULT_EXCLUDE_PATTERNS, ...excludePatterns],
+				useIgnoreFiles: ignoreFiles === false ? undefined : DEFAULT_USE_IGNORE_FILES,
+				useExcludeSettings: excludeSettings ? getExcludeSettingOptions(excludeSettings) : DEFAULT_EXCLUDE_SETTING_OPTIONS,
+			};
+		} else {
+			findOptions = {
+				exclude: undefined,
+				useIgnoreFiles: {
+					local: false,
+					parent: false,
+					global: false,
+				},
+				useExcludeSettings: vscode.ExcludeSettingOptions.None,
+			};
+		}
+
+
+		log.trace(`[${PositronAssistantToolName.ProjectTree}] Constructing project tree with options: ${JSON.stringify({
+			include: filePatterns,
+			exclude: findOptions.exclude,
+			useIgnoreFiles: findOptions.useIgnoreFiles,
+			useExcludeSettings: findOptions.useExcludeSettings,
+			filterResults: filterResultsEnabled,
+			maxFiles: filesLimit,
+		}, null, 2)}`);
+
+		// Construct the project tree
+		const workspaceTrees: DirectoryInfo[] = [];
+		for (const folder of workspaceFolders) {
+			// NOTE: this will not include empty directories :/
+			const matchedFileUris = await vscode.workspace.findFiles2(
+				filePatterns,
+				findOptions,
+				token
+			);
```

Comment:
just want to highlight this callout that empty directories unfortunately are not captured